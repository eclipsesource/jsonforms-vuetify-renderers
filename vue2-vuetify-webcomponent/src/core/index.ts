export * from './dynamicDefaults';
export * from './i18n';
export * from './json-refs';
export * from './keywords';
export * from './transform';
export * from './types';
export * from './validate';
