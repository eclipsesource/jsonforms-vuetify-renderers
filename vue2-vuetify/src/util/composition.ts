import {
  composePaths,
  computeLabel,
  getFirstPrimitiveProp,
  isDescriptionHidden,
  JsonFormsSubStates,
  Resolve,
} from '@jsonforms/core';
import cloneDeep from 'lodash/cloneDeep';
import merge from 'lodash/merge';
import { useStyles } from '../styles';
import { computed, ComputedRef, inject, ref } from '../vue';
import { Ajv } from 'ajv';

const useControlAppliedOptions = <I extends { control: any }>(input: I) => {
  return computed(() =>
    merge(
      {},
      cloneDeep(input.control.value.config),
      cloneDeep(input.control.value.uischema.options)
    )
  );
};

const useComputedLabel = <I extends { control: any }>(
  input: I,
  appliedOptions: ComputedRef<any>
) => {
  return computed((): string => {
    return computeLabel(
      input.control.value.label,
      input.control.value.required,
      !!appliedOptions.value?.hideRequiredAsterisk
    );
  });
};

/**
 * Adds styles, isFocused, appliedOptions and onChange
 */
export const useVuetifyControl = <
  I extends { control: any; handleChange: any }
>(
  input: I,
  adaptValue: (target: any) => any = (v) => v
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const isFocused = ref(false);
  const onChange = (value: any) => {
    input.handleChange(input.control.value.path, adaptValue(value));
  };

  const persistentHint = (): boolean => {
    return !isDescriptionHidden(
      input.control.value.visible,
      input.control.value.description,
      isFocused.value,
      !!appliedOptions.value?.showUnfocusedDescription
    );
  };

  const computedLabel = useComputedLabel(input, appliedOptions);

  const controlWrapper = computed(() => {
    const { id, description, errors, label, visible, required } =
      input.control.value;
    return { id, description, errors, label, visible, required };
  });

  const styles = useStyles(input.control.value.uischema);

  return {
    ...input,
    styles,
    isFocused,
    appliedOptions,
    controlWrapper,
    onChange,
    persistentHint,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyLayout = <I extends { layout: any }>(input: I) => {
  const appliedOptions = computed(() =>
    merge(
      {},
      cloneDeep(input.layout.value.config),
      cloneDeep(input.layout.value.uischema.options)
    )
  );
  return {
    ...input,
    styles: useStyles(input.layout.value.uischema),
    appliedOptions,
  };
};

/**
 * Adds styles, appliedOptions and childUiSchema
 */
export const useVuetifyArrayControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const computedLabel = useComputedLabel(input, appliedOptions);

  const childLabelForIndex = (index: number) => {
    const childLabelProp =
      input.control.value.uischema.options?.childLabelProp ??
      getFirstPrimitiveProp(input.control.value.schema);
    if (!childLabelProp) {
      return `${index}`;
    }
    const labelValue = Resolve.data(
      input.control.value.data,
      composePaths(`${index}`, childLabelProp)
    );
    if (
      labelValue === undefined ||
      labelValue === null ||
      Number.isNaN(labelValue)
    ) {
      return '';
    }
    return `${labelValue}`;
  };
  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
    childLabelForIndex,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyBasicControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
  };
};

/**
 * Extracts Ajv from JSON Forms
 */
export const useAjv = () => {
  const jsonforms = inject<JsonFormsSubStates>('jsonforms');

  if (!jsonforms) {
    throw new Error(
      "'jsonforms' couldn't be injected. Are you within JSON Forms?"
    );
  }

  // should always exist
  return jsonforms.core?.ajv as Ajv;
};
