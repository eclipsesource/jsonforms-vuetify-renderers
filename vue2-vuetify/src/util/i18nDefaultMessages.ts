export const i18nDefaultMessages = {
  arraylayout: {
    add: 'Add',
    delete: 'Delete',
    moveUp: 'Move Up',
    moveDown: 'Move Down',
    dialogTitle: 'Delete {{ element }}?',
    dialogText: 'The element will be deleted.',
    dialogConfirm: 'Delete',
    dialogCancel: 'Cancel',
  },
};
