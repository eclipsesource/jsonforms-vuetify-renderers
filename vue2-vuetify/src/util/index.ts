export * from './options';
export * from './composition';
export * from './validator';
export * from './datejs';
export * from './i18nDefaultMessages';
