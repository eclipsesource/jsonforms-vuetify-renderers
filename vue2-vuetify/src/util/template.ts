import templateFn from 'lodash/template';
import merge from 'lodash/merge';
import { TemplateExecutor, TemplateOptions } from 'lodash';

const interpolate = /{{([\s\S]+?)}}/g;

export const template = (
  templateString?: string,
  options?: TemplateOptions
): TemplateExecutor => {
  return templateFn(
    templateString,
    merge({ interpolate: interpolate }, options)
  );
};
