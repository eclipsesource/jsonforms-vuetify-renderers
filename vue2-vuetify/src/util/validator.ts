import { Options } from 'ajv';
import { createAjv as createAjvCore } from '@jsonforms/core';
import Ajv from 'ajv';

export const createAjv = (options?: Options): Ajv => {
  const ajv = createAjvCore(options);
  ajv.addFormat('password', (_) => true);
  return ajv;
};
