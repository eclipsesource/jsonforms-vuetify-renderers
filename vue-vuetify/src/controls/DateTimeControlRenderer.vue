<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-text-field
      :id="control.id + '-input'"
      :class="styles.control.input"
      :disabled="!control.enabled"
      :autofocus="appliedOptions.focus"
      :placeholder="appliedOptions.placeholder"
      :label="computedLabel"
      :hint="control.description"
      :persistent-hint="persistentHint()"
      :required="control.required"
      :error-messages="control.errors"
      v-bind="vuetifyProps('v-text-field')"
      :model-value="dataTime"
      @update:model-value="onChange"
      @focus="isFocused = true"
      @blur="handleBlur"
      type="datetime-local"
    >
    </v-text-field>
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  isDateTimeControl,
  JsonFormsRendererRegistryEntry,
  rankWith,
} from '@jsonforms/core';
import { defineComponent } from 'vue';
import {
  rendererProps,
  RendererProps,
  useJsonFormsControl,
} from '@jsonforms/vue';
import { VTextField } from 'vuetify/components';
import { useVuetifyControl } from '../util';
import { default as ControlWrapper } from './ControlWrapper.vue';

const toISOString = (inputDateTime: string) => {
  return inputDateTime === '' ? undefined : inputDateTime + ':00.000Z';
};

const controlRenderer = defineComponent({
  name: 'datetime-control-renderer',
  components: {
    ControlWrapper,
    VTextField,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    const adaptValue = (value: any) => toISOString(value) || undefined;

    const control = useVuetifyControl(useJsonFormsControl(props), adaptValue);
    return { ...control, adaptValue };
  },
  computed: {
    dataTime(): string {
      return (this.control.data ?? '').substr(0, 16);
    },
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(2, isDateTimeControl),
};
</script>
