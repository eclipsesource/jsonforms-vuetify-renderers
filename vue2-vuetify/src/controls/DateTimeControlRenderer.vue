<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-text-field
      type="datetime-local"
      :id="control.id + '-input'"
      :class="styles.control.input"
      :disabled="!control.enabled"
      :autofocus="appliedOptions.focus"
      :placeholder="appliedOptions.placeholder"
      :label="computedLabel"
      :hint="control.description"
      :persistent-hint="persistentHint()"
      :required="control.required"
      :error-messages="control.errors"
      :value="dataTime"
      @focus="isFocused = true"
      @blur="isFocused = false"
    />
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  JsonFormsRendererRegistryEntry,
  rankWith,
  isDateTimeControl,
} from '@jsonforms/core';
import { defineComponent } from '../vue';
import {
  rendererProps,
  useJsonFormsControl,
  RendererProps,
} from '@jsonforms/vue2';
import { default as ControlWrapper } from './ControlWrapper.vue';
import { useVuetifyControl, parseDateTime } from '../util';
import { VTextField } from 'vuetify/lib';

const controlRenderer = defineComponent({
  name: 'datetime-control-renderer',
  components: {
    ControlWrapper,
    VTextField,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    return useVuetifyControl(useJsonFormsControl(props));
  },
  computed: {
    dataTime: {
      get(): string | null | undefined {
        const datetimeLocalFormat = 'YYYY-MM-DDTHH:mm:ss.SSS';
        const saveFormat = this.appliedOptions.dateTimeSaveFormat ?? undefined;
        const value = this.control.data as string | undefined | null;

        const dateTime = parseDateTime(value, saveFormat);
        return dateTime ? dateTime.local().format(datetimeLocalFormat) : value;
      },
      set(value: string) {
        const datetimeLocalFormats = [
          'YYYY-MM-DDTHH:mm:ss.SSS',
          'YYYY-MM-DDTHH:mm:ss',
          'YYYY-MM-DDTHH:mm',
        ];
        const saveFormat =
          this.appliedOptions.dateTimeSaveFormat ?? 'YYYY-MM-DDTHH:mm:ssZ';

        const dateTime = parseDateTime(value, datetimeLocalFormats);
        const result = dateTime ? dateTime.format(saveFormat) : value;

        this.onChange(result);
      },
    },
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(2, isDateTimeControl),
};
</script>
