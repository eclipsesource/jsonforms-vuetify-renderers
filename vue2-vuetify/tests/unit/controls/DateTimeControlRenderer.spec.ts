import { clearAllIds } from '@jsonforms/core';
import { Wrapper } from '@vue/test-utils';
import DateTimeControlRenderer, {
  entry as dateTimeControlRendererEntry,
} from '../../../src/controls/DateTimeControlRenderer.vue';
import { wait } from '../../../tests';
import { mountJsonForms } from '../util';

describe('DateTimeControlRenderer.vue', () => {
  const renderers = [dateTimeControlRendererEntry];

  const data = '2021-03-09T21:54:00.000Z';
  const schema = {
    type: 'string',
    title: 'My Date Time',
    format: 'date-time',
  };
  const uischema = {
    type: 'Control',
    scope: '#',
    options: {
      placeholder: 'date-time placeholder',
    },
  };
  let wrapper: Wrapper<any, Element>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('check if child DateTimeControlRenderer exists', () => {
    expect(wrapper.getComponent(DateTimeControlRenderer));
  });

  it('renders a date time input', () => {
    expect(wrapper.find('input[type="datetime-local"]').exists()).toBe(true);
  });

  it('renders title as label', () => {
    expect(wrapper.find('label').text()).toEqual('My Date Time');
  });

  it('emits a data change', async () => {
    const input = wrapper.find('input[type="datetime-local"]');
    await input.setValue('2021-03-10T21:10');
    await input.trigger('blur');
    // 300 ms debounceWait
    await wait(300);
    expect(wrapper.vm.$data.data).toEqual('2021-03-10T21:10:00.000Z');
  });

  it('should have a placeholder', async () => {
    const input = wrapper.find('input[type="datetime-local"]');
    // select the input so the placeholder is generated
    await input.trigger('click');
    const placeholder = input.attributes('placeholder');
    expect(placeholder).toEqual('date-time placeholder');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
