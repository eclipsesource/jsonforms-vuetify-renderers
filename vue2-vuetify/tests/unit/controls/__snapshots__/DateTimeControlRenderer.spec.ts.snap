// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeControlRenderer.vue should render component and match snapshot 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My Date Time"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--placeholder input">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light"
                     style="left: 0px; position: absolute;"
              >
                My Date Time
              </label>
              <input id="#-input"
                     type="text"
              >
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-menu">
      </div>
    </div>
  </div>
</div>
`;

exports[`DateTimeControlRenderer.vue should render component and match snapshot when clicked 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My Date Time"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--placeholder input">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light"
                     style="left: 0px; position: absolute;"
              >
                My Date Time
              </label>
              <input id="#-input"
                     type="text"
              >
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-menu">
        <transition-stub name="scale-transition">
        </transition-stub>
      </div>
    </div>
  </div>
  <div role="menu"
       class="v-menu__content theme--light menuable__content__active "
       style="max-height: auto; min-width: 580px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 8; display: none;"
  >
    <div class="v-card v-sheet theme--light">
      <div class="row no-gutters">
        <div min-width="290px"
             class="col col-auto"
        >
          <div class="v-picker v-card v-picker--date theme--light">
            <div class="v-picker__title primary">
              <div class="v-date-picker-title">
                <div class="v-picker__title__btn v-date-picker-title__year">
                  2021
                </div>
                <div class="v-picker__title__btn v-date-picker-title__date v-picker__title__btn--active">
                  <transition-stub name="picker-transition">
                    <div>
                      Tue, Mar 9
                    </div>
                  </transition-stub>
                </div>
              </div>
            </div>
            <div class="v-picker__body theme--light"
                 style="width: 290px;"
            >
              <transition-stub name="fade-transition">
                <div>
                  <div class="v-date-picker-header theme--light">
                    <button type="button"
                            class="v-btn v-btn--icon v-btn--round theme--light v-size--default"
                            aria-label="Previous month"
                    >
                      <span class="v-btn__content">
                        <i aria-hidden="true"
                           class="v-icon notranslate mdi mdi-chevron-left theme--light"
                        >
                        </i>
                      </span>
                    </button>
                    <div class="v-date-picker-header__value">
                      <transition-stub name="tab-transition">
                        <div class="accent--text">
                          <button type="button">
                            March 2021
                          </button>
                        </div>
                      </transition-stub>
                    </div>
                    <button type="button"
                            class="v-btn v-btn--icon v-btn--round theme--light v-size--default"
                            aria-label="Next month"
                    >
                      <span class="v-btn__content">
                        <i aria-hidden="true"
                           class="v-icon notranslate mdi mdi-chevron-right theme--light"
                        >
                        </i>
                      </span>
                    </button>
                  </div>
                  <div class="v-date-picker-table v-date-picker-table--date theme--light">
                    <transition-stub name="tab-transition">
                      <table>
                        <thead>
                          <tr>
                            <th>
                              S
                            </th>
                            <th>
                              M
                            </th>
                            <th>
                              T
                            </th>
                            <th>
                              W
                            </th>
                            <th>
                              T
                            </th>
                            <th>
                              F
                            </th>
                            <th>
                              S
                            </th>
                          </tr>
                        </thead>
                        <tbody>
                          <tr>
                            <td>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  1
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  2
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  3
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  4
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  5
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  6
                                </div>
                              </button>
                            </td>
                          </tr>
                          <tr>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  7
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  8
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--active v-btn--rounded theme--light accent"
                              >
                                <div class="v-btn__content">
                                  9
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  10
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  11
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  12
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  13
                                </div>
                              </button>
                            </td>
                          </tr>
                          <tr>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  14
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  15
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  16
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  17
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  18
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  19
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  20
                                </div>
                              </button>
                            </td>
                          </tr>
                          <tr>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  21
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  22
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  23
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  24
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  25
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  26
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  27
                                </div>
                              </button>
                            </td>
                          </tr>
                          <tr>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  28
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  29
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  30
                                </div>
                              </button>
                            </td>
                            <td>
                              <button type="button"
                                      class="v-btn v-btn--text v-btn--rounded theme--light"
                              >
                                <div class="v-btn__content">
                                  31
                                </div>
                              </button>
                            </td>
                            <td>
                            </td>
                            <td>
                            </td>
                            <td>
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </transition-stub>
                  </div>
                </div>
              </transition-stub>
            </div>
          </div>
        </div>
        <div min-width="290px"
             class="col col-auto"
        >
          <div class="v-picker v-card v-picker--time theme--light">
            <div class="v-picker__title primary">
              <div class="v-time-picker-title">
                <div class="v-time-picker-title__time">
                  <div class="v-picker__title__btn v-picker__title__btn--active">
                    21
                  </div>
                  <span>
                    :
                  </span>
                  <div class="v-picker__title__btn">
                    54
                  </div>
                </div>
              </div>
            </div>
            <div class="v-picker__body theme--light"
                 style="width: 290px;"
            >
              <transition-stub name="fade-transition">
                <div class="v-time-picker-clock__container">
                  <div class="v-time-picker-clock theme--light">
                    <div class="v-time-picker-clock__inner">
                      <div class="v-time-picker-clock__hand v-time-picker-clock__hand--inner accent"
                           style="transform: rotate(630deg) scaleY(0.62);"
                      >
                      </div>
                      <span class="v-time-picker-clock__item"
                            style="left: 50%; top: 0%;"
                      >
                        <span>
                          0
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 75%; top: 6.698729810778062%;"
                      >
                        <span>
                          1
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 93.30127018922192%; top: 24.999999999999993%;"
                      >
                        <span>
                          2
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 100%; top: 50%;"
                      >
                        <span>
                          3
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 93.30127018922194%; top: 74.99999999999999%;"
                      >
                        <span>
                          4
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 75.00000000000001%; top: 93.30127018922192%;"
                      >
                        <span>
                          5
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 50.00000000000001%; top: 100%;"
                      >
                        <span>
                          6
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 25.000000000000014%; top: 93.30127018922194%;"
                      >
                        <span>
                          7
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 6.698729810778076%; top: 75.00000000000003%;"
                      >
                        <span>
                          8
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 0%; top: 50.00000000000001%;"
                      >
                        <span>
                          9
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 6.6987298107780475%; top: 25.000000000000032%;"
                      >
                        <span>
                          10
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 24.99999999999998%; top: 6.698729810778083%;"
                      >
                        <span>
                          11
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 49.99999999999999%; top: 19%;"
                      >
                        <span>
                          12
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 65.49999999999997%; top: 23.153212482682388%;"
                      >
                        <span>
                          13
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 76.84678751731758%; top: 34.499999999999986%;"
                      >
                        <span>
                          14
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 81%; top: 49.999999999999964%;"
                      >
                        <span>
                          15
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 76.84678751731761%; top: 65.49999999999997%;"
                      >
                        <span>
                          16
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 65.50000000000004%; top: 76.84678751731758%;"
                      >
                        <span>
                          17
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 50.000000000000014%; top: 81%;"
                      >
                        <span>
                          18
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 34.50000000000003%; top: 76.84678751731761%;"
                      >
                        <span>
                          19
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 23.153212482682427%; top: 65.50000000000004%;"
                      >
                        <span>
                          20
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item v-time-picker-clock__item--active accent"
                            style="left: 19%; top: 50.000000000000014%;"
                      >
                        <span>
                          21
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 23.153212482682385%; top: 34.50000000000003%;"
                      >
                        <span>
                          22
                        </span>
                      </span>
                      <span class="v-time-picker-clock__item"
                            style="left: 34.49999999999996%; top: 23.153212482682427%;"
                      >
                        <span>
                          23
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <button type="button"
                class="v-btn v-btn--text theme--light v-size--default"
        >
          <span class="v-btn__content">
            Clear
          </span>
        </button>
        <div class="spacer">
        </div>
        <button type="button"
                class="v-btn v-btn--text theme--light v-size--default"
        >
          <span class="v-btn__content">
            Cancel
          </span>
        </button>
        <button type="button"
                class="v-btn v-btn--text theme--light v-size--default primary--text"
        >
          <span class="v-btn__content">
            OK
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
